text/002x-succinct-atomic-swaps.md

```diff
@@ -0,0 +1,139 @@
+
+The reference implementation uses Bitcoin as the other chain in the swap, but potentially any other chain using Secp256k1 keys and multisignatures can work with the protocol.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Funds on the non-Grin chain need to be locked in a multisignature transaction that requires knowledge of both "atomic nonces" to be spent. A refund transaction is also needed to recover funds in case one of the parties becomes non-responsive, or the protocol needs to be aborted for another reason.
+
+## Atomic nonces
+[atomic-nonces]: #atomic-nonces
+
+In the case of Grin, atomic nonces are just Secp256k1 private keys. Let's say Alice knows atomic nonce `aN`, and Bob knows atomic nonce `bN`.
+
+In the refund transaction, Alice (with locked Grin funds) will reveal her nonce `aN` to allow Bob to recover his locked funds on the other chain.
+
+In the main transaction, Bob and Alice create a multisignature Grin transaction which reveals Bob's nonce `bN` to Alice. Alice can then recover the funds locked on the other chain.
+
+Atomic IDs are identifiers used to derive atomic nonces using the keychain. Atomic IDs are formatted as follows:
+
+```
+0x4 | 'mwatomic' | 0x0 0x0 0x0 0x0 | 32-bit ID
+```
+
+The 32-bit ID can be expanded to 64-bits if necessary, and doesn't need to be random (it can be a simple counter). However, an ID can only be used once.
+
+Atomic nonces must only be used **once**. Reuse potentially results in loss of funds, if the other party has the previously used nonce.
+
+Without automated defenses, a malicious party could trick a user into deriving the same nonce (by supplying a duplicate atomic ID). The malicious user could then recover funds on the other chain without completing the atomic swap protocol.
+
+Currently, the automated defense is to use a space-efficient filter (`scalable cuckoo filter`), which is stored in the wallet's database. The filter cannot be reconstructed if the wallet file is deleted and recovered, so it is recommended to use a new keymask for any future atomic swaps.
+
+## Adaptor signatures
+[adaptor-signatures]: #adaptor-signatures
+
+Adaptor signatures were originally invented by Andrew Poelstra [\[3\]](#references). Adaptor signatures use additive properties of Schnorr signatures to commit to a secret, then later reveal it by creating a full Schnorr signature over the same message.
+
+
+## Protocol rounds
+[protocol-rounds]: #protocol-rounds
+
```

Comment:
Would be nice to have a picture here, as Somsen provided. How about

```Grin-BTC Succinct Atomic Swap

+----------+        +-------------+  1 day   +-------------+  2 days     +----------+
| 87k Grin |  --->  | 87k Grin    |  ----->  | 87k Grin    |  -------->  | 87k Grin |
| Alice    |        | Alice + Bob |          | Alice + Bob |  reveals    | Alice    |
+----------+        +-------------+          +-------------+  secretA    +----------+
                                                       
                            |                       | 3
                            |                       | days
                            |                       v
                            |                         
                            |                +----------+
                            +------------->  | 87k Grin |
                             reveals         | Bob      |
                             secretB         +----------+



+-------+         +-------------------+
| 1 BTC |  ---->  | 1 BTC             |
| Bob   |         | secretA + secretB |
+-------+         +-------------------+
```

Reply:
Thanks for the ASCII art! This is great, added to the section.